lang/java/avro/src/main/java/org/apache/avro/io/parsing/ResolvingGrammarGenerator.java

```diff
@@ -294,7 +296,14 @@ static void encode(Encoder e, Schema s, JsonNode n) throws IOException {
         throw new AvroTypeException("Non-string default value for fixed: " + n);
       byte[] bb = n.textValue().getBytes(StandardCharsets.ISO_8859_1);
       if (bb.length != s.getFixedSize()) {
-        bb = Arrays.copyOf(bb, s.getFixedSize());
+        int offset = s.getFixedSize() - bb.length;
+        byte signum = (byte) (new BigDecimal(new BigInteger(bb)).signum() < 0 ? 0xFF : 0x00);
```

Comment:
I checked that this works for negative default numbers!  

Is checking `...signum() < 0` the same as checking whether `bb[0] < 0` ?

I'm also tempted to say that if the length of the default is smaller than the specified fixed size, it should be positive.

Reply:
Hi, for example the decimal number "-34.34", the padding should be negative
I think in Conversions, the handling of size mismatch is very good
[](https://github.com/apache/avro/blob/29db1adf5b9052d6d57a5c9f9e855a44e7c8d84b/lang/java/avro/src/main/java/org/apache/avro/Conversions.java#L109)